### ISSUE TITLE
Port is not optional in URI map

### ISSUE THREAD

**Issue by lvh** [babashka/http-client#73]
Works as expected:

```
(babashka.http-client.internal.helpers/->uri
 {:host "foo.com" :port 443 :scheme "https" :path "/"})
```

Breaks (note missing port):

```
(babashka.http-client.internal.helpers/->uri
 {:host "foo.com" :scheme "https" :path "/"})
```

With following incomprehensible error:

```
;; 1. Unhandled java.lang.NullPointerException
;; Cannot invoke "java.lang.Character.charValue()" because "x" is null
```

Java environment:

```
(System/getProperty "java.version")
;; => "21.0.4"

(System/getProperty "java.vm.vendor")
;; => "Eclipse Adoptium"
```

I believe this is because the port is specified as a primitive. The correct value to set is -1 for the implicit default value. This code assumes nil is okay; it is not.

This behavior contradicts the documentation for `request` (which marks `:host` as required but not `:port`), though I believe the right approach isn't to amend the docs but rather add the -1 behavior I mentioned above.

**Comment by lvh** [comment]
Attached PR fixes the issue.

**Comment by borkdude** [comment]
@lvh Thanks! And it was nice meeting you (again) at Heart of Clojure

I'd like to learn more about this use case. The namespace marked as internal isn't for public use, it may change at any time.
Did you run into this issue using any of the public APIs? And is there any documentation regarding setting ports to -1 somewhere as a best practice? I'm unfamiliar with this.

**Comment by lvh** [comment]
@borkdude Likewise!

> Did you run into this issue using any of the public APIs?

Yes, passing a `:uri` map to the main `request` API entrypoint caused the bug.

> And is there any documentation regarding setting ports to -1 somewhere as a best practice?

Yes, I'd say that it's more than just "best practice" -- it's not some general convention (though port of -1 does show up in different contexts for similar reasons, i.e. having a non-nillable type), it's specifically how this class is meant to be used. From the [Javadoc](https://docs.oracle.com/javase/8/docs/api/index.html?java/net/URI.html):

> In a given instance any particular component is either undefined or defined with a distinct value. Undefined string components are represented by null, while undefined integer components are represented by -1.

**Comment by lvh** [comment]
(Additionally, please note that `request` 's docstring suggests the URI can be a map and the port is optional, which isn't true right now.)

**Comment by borkdude** [comment]
Thanks man! Is it possible to write a test for this using the public APIs? If not, no big deal, but usually I keep my tests limited to the public APIs

**Comment by lvh** [comment]
I tried writing such a test (actually specifically I tried extending the closest current test: https://github.com/babashka/http-client/blob/efef89605f2e119dd9bf98b845b1a471b26cd4e9/test/babashka/http_client_test.clj#L251) but there was no obvious way to do that. The default ports are all privileged, so the current approach of using a real web server doesn't work.

I guess I could technically create a fake client that does nothing and test that the request call doesn't throw...? It looks like the  tests already make real requests to public websites so I guess I could make a real request to github.com?
